components/Feed/Caption.tsx

```diff
@@ -0,0 +1,45 @@
+"use client";
+import { useEffect, useState } from "react";
+
+export default function Caption({ caption }: { caption: string | null }) {
+  // track if the caption is truncated
+  const [isCaptionTruncated, setIsCaptionTruncated] = useState<boolean>(false);
+  // caption shown on the video slide
+  const [displayCaption, setDisplayCaption] = useState<string | null>(caption);
+  // track if the caption is expanded
+  const [isExpanded, setIsExpanded] = useState<boolean>(false);
+
+  const truncationCharacterLimit = 80;
+
+  // truncate the caption if it's is greater than truncationCharacterLimit
+  useEffect(() => {
+    if (caption && caption.length > 100) {
+      setIsCaptionTruncated(true);
+      setDisplayCaption(caption.slice(0, truncationCharacterLimit) + "...");
+    }
+  }, [caption]);
+
+  // toggle expansion state when the "more" or "less" button is clicked
+  const handleExpansion = () => {
+    setIsExpanded(!isExpanded);
+    if (isExpanded) {
+      setDisplayCaption(caption?.slice(0, truncationCharacterLimit) + "...");
+    } else {
+      setDisplayCaption(caption);
+    }
+  };
+
+  return (
+    <div>
+      <p className="inline">{displayCaption}</p>
+      {isCaptionTruncated && (
+        <>
+          {" "}
+          <button onClick={handleExpansion} className="text-white/80">
+            {isExpanded ? "less" : "more"}
+          </button>
+        </>
+      )}
+    </div>
+  );
```

Comment:
An alternative is to do the truncation on the fly, which simplifies the code a little since you can remove some of the states:
```ts
  const TRUNCATION_LIMIT = 80;
  const shouldTruncate = caption.length > TRUNCATION_LIMIT + 20;
  const [isTruncated, setIsTruncated] = useState<boolean>(shouldTruncate);

  return (
    <p>
      {isTruncated ? `${caption.slice(0, TRUNCATION_LIMIT)}...` : caption}
      {shouldTruncate ? (
        <>
          {" "}
          <button
            onClick={() => setIsTruncated(!isTruncated)}
            className="text-white/80 hover:text-white"
          >
            {isTruncated ? "more" : "less"}
          </button>
        </>
      ) : null}
    </p>
  );
```